contrib/cloud.google.com/go/pubsub.v1/pubsub.go

```diff
@@ -0,0 +1,87 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// Package pubsub provides functions to trace the cloud.google.com/pubsub/go package.
+package pubsub
+
+import (
+	"context"
+	"sync"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+
+	"cloud.google.com/go/pubsub"
+)
+
+// Publish publishes a message on the specified topic and returns a PublishResult.
+// This function is functionally equivalent to t.Publish(ctx, msg), but it also starts a publish
+// span and it ensures that the datadog tracing metadata is propagated as attributes attached to
+// the published message.
+func Publish(ctx context.Context, t *pubsub.Topic, msg *pubsub.Message) *PublishResult {
+	span, ctx := tracer.StartSpanFromContext(
+		ctx,
+		"pubsub.publish",
+		tracer.ResourceName(t.String()),
+		tracer.SpanType(ext.SpanTypeMessageProducer),
+		tracer.Tag("message_size", len(msg.Data)),
+		tracer.Tag("ordering_key", msg.OrderingKey),
+	)
+	if msg.Attributes == nil {
+		msg.Attributes = make(map[string]string)
+	}
+	if err := tracer.Inject(span.Context(), tracer.TextMapCarrier(msg.Attributes)); err != nil {
+		log.Debug("contrib/cloud.google.com/go/pubsub.v1/: failed injecting tracing attributes: %v", err)
+	}
+	span.SetTag("num_attributes", len(msg.Attributes))
+	return &PublishResult{
+		PublishResult: t.Publish(ctx, msg),
+		span:          span,
+	}
+}
+
+// PublishResult wraps *pubsub.PublishResult
+type PublishResult struct {
+	*pubsub.PublishResult
+	once sync.Once
+	span tracer.Span
+}
+
+// Get wraps (pubsub.PublishResult).Get(ctx). When this function returns the publish span
+// created in Publish is completed.
+func (r *PublishResult) Get(ctx context.Context) (string, error) {
+	serverID, err := r.PublishResult.Get(ctx)
+	r.once.Do(func() {
```

Comment:
I have never used this library (pubsub). Is this common pattern that once a message is published a get follows shortly after? Is it possible that someone never calls Get leaving us with a span opened forever? If the latter is a possibility it might create issues.

Also, we can test it, but it might be fine to drop the `sync.Once` - I think `Finish` is already idempotent.

Reply:
> I have never used this library (pubsub). Is this common pattern that once a message is published a get follows shortly after? 

We do, as otherwise it's impossible to know if the request succeeded or not. 

> Is it possible that someone never calls Get leaving us with a span opened forever? If the latter is a possibility it might create issues.

I think you're not supposed to leave it hanging, but I can't be 100% sure nobody is doing this.

How do we handle the case of someone not calling Finish() on a span?

I could do something like using a background worker, or spawning a goroutine per request, or something unspeakable with finalizers... but all of these options will complicate things significantly.

> Also, we can test it, but it might be fine to drop the `sync.Once` - I think `Finish` is already idempotent.

Oh ok, wasn't sure if this was guaranteed. What about concurrent calls?

Reply:
> I could do something like using a background reaper, or spawning a goroutine per request, or something unspeakable with finalizers... but all of these options will complicate things significantly.

It's clear now. Please don't 😄 

> Oh ok, wasn't sure if this was guaranteed. What about concurrent calls?

It would work in all cases, but the problem is that `span.SetTag` has a lock so if you were to remove `sync.Once` it could add lock contention, which you may not want. So let's leave it as it is.

Resolving.